scipy/optimize/least_squares.py

```diff
@@ -0,0 +1,657 @@
+"""Generic interface for least-square minimization."""
+
+from warnings import warn
+
+import numpy as np
+from numpy.linalg import norm
+
+from . import _minpack
+from .optimize import OptimizeResult
+from ._numdiff import approx_derivative, group_columns
+from ..sparse import issparse, csr_matrix, csc_matrix
+from ..sparse.linalg import LinearOperator
+
+from ._lsq_common import in_bounds, prepare_bounds
+from ._lsq_trf import trf
+from ._lsq_dogbox import dogbox
+
+__all__ = ['least_squares']
+
+
+EPS = np.finfo(float).eps
+
+
+def check_tolerance(ftol, xtol, gtol):
+    message = "{} is too low, setting to machine epsilon {}."
+    if ftol < EPS:
+        warn(message.format("`ftol`", EPS))
+        ftol = EPS
+    if xtol < EPS:
+        warn(message.format("`xtol`", EPS))
+        xtol = EPS
+    if gtol < EPS:
+        warn(message.format("`gtol`", EPS))
+        gtol = EPS
+
+    return ftol, xtol, gtol
+
+
+TERMINATION_MESSAGES = {
+    -1: "Improper input parameters status returned from `leastsq`",
+    0: "The maximum number of function evaluations is exceeded.",
+    1: "`gtol` termination condition is satisfied.",
+    2: "`ftol` termination condition is satisfied.",
+    3: "`xtol` termination condition is satisfied.",
+    4: "Both `ftol` and `xtol` termination conditions are satisfied."
+}
+
+
+FROM_MINPACK_TO_COMMON = {
+    0: -1,  # 0 improper input parameters for MINPACK.
+    1: 2,
+    2: 3,
+    3: 4,
+    4: 1,
+    5: 0
+    # There are 6, 7, 8 for too small tolerance parameters,
+    # but we guard against them by checking ftol, xtol, gtol beforehand.
+}
+
+
+def call_minpack(fun, x0, jac, ftol, xtol, gtol, max_nfev, scaling, diff_step):
+    n = x0.size
+
+    if diff_step is None:
+        epsfcn = np.finfo(float).eps
```

Comment:
Nitpick: The call to finfo(float) is repeated at least four times, twice at module level. Better do it once, store the result, and import.